File: wheeled_vehicle_models/18dof/demos/HMMWV/demo_hmmwv_hi_stepWithJac.cpp
#include <numeric>
#include <algorithm>
#include <iterator>
#include <chrono>

#include "dof18_halfImplicit.h"

using std::chrono::high_resolution_clock;
using namespace d18;

void printVectorOfVectors(const std::vector<std::vector<double>>& vec) {
    for (const auto& inner_vec : vec) {
        for (const auto& element : inner_vec) {
            std::cout << element << " ";
        }
        std::cout << std::endl;
    }
}

int main(int argc, char** argv) {
    // Driver inputs and reference trajectory
    std::string driver_file = "../data/input/" + std::string(argv[1]) + ".txt";

    // Vehicle specification
    std::string vehParamsJSON = (char*)"../data/json/HMMWV/vehicle.json";
    std::string tireParamsJSON = (char*)"../data/json/HMMWV/tmeasy.json";

    // Construct the solver
    d18SolverHalfImplicit solver;
    solver.Construct(vehParamsJSON, tireParamsJSON, driver_file);

    // Set time step
    solver.SetTimeStep(1e-3);

    // Initialize solver (set initial conditions)
    VehicleState veh_st;
    TMeasyState tirelf_st;
    TMeasyState tirerf_st;
    TMeasyState tirelr_st;
    TMeasyState tirerr_st;
    solver.Initialize(veh_st, tirelf_st, tirerf_st, tirelr_st, tirerr_st);

    // Enable output
    solver.SetOutput("../data/output/" + std::string(argv[1]) + "_hmmwv18HiStep.csv", 100);

    double timeStep = solver.GetStep();
    double endTime = solver.GetEndT();
    double t = 0;
    double new_time = 0;
    DriverData driver_data = solver.GetDriverData();  // Get the driver data associated with input file

    // Time the while loop
    auto start = high_resolution_clock::now();

    // Start the loop
    while (t < (endTime - timeStep / 10.)) {
        auto controls = GetDriverInput(t, driver_data);  // Get the controls for the current time
        new_time =
            solver.IntegrateStepWithJacobian(t, controls.m_throttle, controls.m_steering, controls.m_braking, true);

// Print the Jacobian state and controls
#ifdef DEBUG
        std::cout << "Jacobian state: " << std::endl;
        printVectorOfVectors(solver.GetJacobianState());
        std::cout << "Jacobian controls: " << std::endl;
        printVectorOfVectors(solver.GetJacobianControls());
        std::cout << std::endl;
#endif

        t = new_time;
    }

    auto stop = high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(stop - start);
    // Print time in milli seconds
    std::cout << "Time taken by function: " << duration.count() / 1000.0 << " milliseconds" << std::endl;
}

File: wheeled_vehicle_models/18dof/pydof18_demos/HMMWV/demo_hmmwv_hi_stepWithJac.py
# =============================================================================
# Authors: Huzaifa Unjhawala
# =============================================================================
#
# This demo is intended to recreate the demo_hmmwv_hi_stepWithJac.cpp demo in python
# In addition, we also show that although we don't necessarily need to use the a driver file
# to run the simulation, we can still optionally use the driver file to get the driver input.
#
# =============================================================================
import time
import numpy as np
import pydof18 as dof18
import sys

# Parameters path
vehParamsJsonPath = "../../data/json/HMMWV/vehicle.json"
tireParamsJsonPath = "../../data/json/HMMWV/tmeasy.json"

# Input file
driver_file = "../../data/input/" + str(sys.argv[1]) + ".txt"

# Construct the solver
solver = dof18.d18SolverHalfImplicit()
solver.Construct(vehParamsJsonPath, tireParamsJsonPath, driver_file)

# Set the simulation time step
solver.SetTimeStep(1e-3)

# Set the initial conditions
veh_st = dof18.VehicleState()
tirelf_st = dof18.TMeasyState()
tirerf_st = dof18.TMeasyState()
tirelr_st = dof18.TMeasyState()
tirerr_st = dof18.TMeasyState()
solver.Initialize(veh_st, tirelf_st, tirerf_st, tirelr_st, tirerr_st)


timeStep = solver.GetStep()
endTime = solver.GetEndT()
t = 0
new_time = 0

driver_data = solver.GetDriverData()

# time this loop
start = time.time()
while (t < (endTime - timeStep / 10.)):
    controls = dof18.GetDriverInput(t, driver_data)
    new_time = solver.IntegrateStepWithJacobian(
        t, controls.m_throttle, controls.m_steering, 0, True)
    # state_jac = np.array(solver.GetJacobianState())
    # print(state_jac)

    t = new_time

end = time.time()
# Print time in ms
print("Time taken: ", (end - start) * 1000)
